draft-irtf-cfrg-hash-to-curve.md

```diff
@@ -1581,75 +1589,139 @@ Steps:
+k MUST be chosen such that sgn0(k) == 1.
+
+The 4-isogeny map from curve448 to edwards448 ({{RFC7748}}, Section 4.2)
+is unambiguous with respect to sign.
+
+When defining new twisted Edwards curves, a Montgomery equivalent and rational
+map SHOULD be specified, and the sign of the rational map SHOULD be stated
+unambiguously.
+
+When hashing to a twisted Edwards curve that does not have a standardized
+Montgomery form or rational map, the following procedure MUST be
+used to derive them.
 For a twisted Edwards curve given by a * x^2 + y^2 = 1 + d * x^2 * y^2,
-first compute A and B, the parameters of the equivalent Montgomery curve,
-as follows:
+first compute A and B, the parameters of the equivalent curve given by
+y'^2 = x'^3 + A * x'^2 + B * x', as follows:
 
-- A = (a + 2) / 2
+- A = (a + d) / 2
 - B = (a - d)^2 / 16
 
-Next, use A and B as the curve parameters in the Elligator 2 method of
-{{elligator2}} to obtain a point (x', y') on the Montgomery curve.
-Finally, convert (x', y') to a point (x, y) on the target curve.
-Letting B' = 4 / (a - d), compute
+Note that the above curve is given in the Weierstrass form required
+by the Elligator 2 mapping.
+The rational map from the point (x', y') on this Weierstrass curve
+to the point (x, y) on the twisted Edwards curve is given by
 
 - x = x' / y'
-- y = (B' * x' - 1) / (B' * x' + 1)
-
-This can be done in one inversion using Montgomery's trick {{M87}}:
-invert the product y' * (B' * x' + 1), then multiply by y' to obtain
-1 / (B' * x' + 1), and likewise for 1 / y'.
+- y = (B' * x' - 1) / (B' * x' + 1), where B' = 1 / sqrt(B) = 4 / (a - d)
+
+For completeness, we give the inverse map in {{rational-map-inverse}}.
+Note that the inverse map is not used when hashing to a twisted Edwards curve.
+
+Rational maps may be undefined, for example, when the denominator of one
+of the rational functions is zero.
+For example, in the map described above, the exceptional cases are
+y' == 0 or B' * x' == -1.
+Implementations MUST detect exceptional cases and return the value
+(x, y) = (0, 1), which is a valid point on all twisted Edwards curves
+given by the equation above.
+
+The following straight-line implementation of the above rational map
+handles the exceptional cases.
+Implementations of other rational maps (e.g., the ones give in {{RFC7748}})
+are analogous.
+
+~~~
+rational_map(x', y')
+Input: (x', y'), a point on the curve y'^2 = x'^3 + A * x'^2 + B * x'.
+Output: (x, y), a point on the equivalent twisted Edwards curve.
+
+1. t1 = y' * B'
+2. t2 = x' + 1
+3. t3 = t1 * t2
+4. t3 = inv0(t3)
+5.  x = t2 * t3
+6.  x = x * x'
+7.  y = x' - 1
+8.  y = y * t3
+9.  y = y * t1
+10. e = y == 0
+11. y = CMOV(y, 1, e)
+12. return (x, y)
+~~~
 
-Preconditions: A twisted Edwards curve.
+### Elligator 2 Method {#ell2edwards}
 
-Constants:
+Preconditions: A twisted Edwards curve E and an equivalent curve M, either
+in Montgomery form ({{montgomery}}) or in the Weierstrass form given in {{rational-map}}.
```

Comment:
the input of `rational_map` function is always a point in Weierstrass form, so M cannot be in Montgomery form.

Reply:
This isn't quite true: when using a standardized rational map (as in curve25519 or curve448), the rational map is from Montgomery to Edwards, not from Weierstrass to Edwards.

I've simplified the sentence in a way that (I think) addresses your concern.